java-checks/src/main/java/org/sonar/java/checks/ArrayCopyLoopCheck.java

```diff
@@ -0,0 +1,301 @@
+      if (body.size() == 2) {
+        StatementTree update = body.get(1);
+        Symbol counter = checkUpdate(update);
+        if (counter != null) {
+          ExpressionTree condition = tree.condition();
+          if (checkCondition(condition, counter)) {
+            StatementTree statement = body.get(0);
+            return checkStatement(statement, counter);
+          }
+        }
+      }
+    }
+    return false;
+  }
+
+  private static boolean checkForEach(ForEachStatement tree) {
+    ExpressionTree expression = tree.expression();
+    if (isArray(expression)) {
+      StatementTree statement = getStatement(tree);
+      if (statement != null && statement.is(Kind.EXPRESSION_STATEMENT)) {
+        expression = ((ExpressionStatementTree) statement).expression();
+        return isArrayToListCopy(expression, tree.variable());
+      }
+    }
+    return false;
+  }
+
+  private static boolean checkCondition(ExpressionTree tree, Symbol counter) {
+    if (tree.is(Kind.LESS_THAN, Kind.LESS_THAN_OR_EQUAL_TO, Kind.GREATER_THAN, Kind.GREATER_THAN_OR_EQUAL_TO, Kind.NOT_EQUAL_TO)) {
+      BinaryExpressionTree comparison = (BinaryExpressionTree) tree;
+      ExpressionTree lhs = comparison.leftOperand();
+      ExpressionTree rhs = comparison.rightOperand();
+      return isCounter(lhs, counter) ^ isCounter(rhs, counter);
+    }
+    return false;
+  }
+
+  @CheckForNull
+  private static Symbol checkUpdate(StatementTree tree) {
+    Symbol counter = null;
+    if (tree.is(Kind.EXPRESSION_STATEMENT)) {
+      ExpressionStatementTree update = (ExpressionStatementTree) tree;
+      counter = isIncrement(update);
+      if (counter == null) {
+        counter = isPlusAssignment(update);
+        if (counter == null) {
+          counter = isAssignment(update);
+        }
+      }
+    }
+    return counter;
+  }
+
+  private static boolean checkStatement(StatementTree tree, Symbol counter) {
+    if (tree.is(Kind.EXPRESSION_STATEMENT)) {
+      ExpressionTree expression = ((ExpressionStatementTree) tree).expression();
+      return isArrayToArrayCopy(expression, counter) || isArrayToListCopy(expression, counter);
+    }
+    return false;
+  }
+
+  @CheckForNull
+  private static Symbol isIncrement(ExpressionStatementTree update) {
+    if (update.expression().is(Kind.POSTFIX_INCREMENT, Kind.PREFIX_INCREMENT)) {
+      UnaryExpressionTree increment = (UnaryExpressionTree) update.expression();
+      IdentifierTree identifier = (IdentifierTree) increment.expression();
+      return identifier.symbol();
+    }
+    return null;
+  }
+
+  @CheckForNull
+  private static Symbol isPlusAssignment(ExpressionStatementTree update) {
+    if (update.expression().is(Kind.PLUS_ASSIGNMENT)) {
+      AssignmentExpressionTree assignment = (AssignmentExpressionTree) update.expression();
+      ExpressionTree lhs = assignment.variable();
+      ExpressionTree rhs = assignment.expression();
+      IdentifierTree identifier = (IdentifierTree) lhs;
```

Comment:
This cast is not safe, `lhs` can be `MemberSelectExpressionTree` when you use field like
```
    for (this.x = 0; this.x < 10; this.x += 1) {

    }
```